src/app/components/chat/window/output/AppChatWindowOutputItem.vue

```diff
@@ -0,0 +1,530 @@
+import { DefaultTheme } from '../../../../../_common/theme/theme.model';
+import { useThemeStore } from '../../../../../_common/theme/theme.store';
+import { vAppTooltip } from '../../../../../_common/tooltip/tooltip-directive';
+import AppTranslate from '../../../../../_common/translate/AppTranslate.vue';
+import { $gettext } from '../../../../../_common/translate/translate.service';
+import { ChatStore, ChatStoreKey } from '../../chat-store';
+import {
+	removeMessage as chatRemoveMessage,
+	setMessageEditing,
+	userCanModerateOtherUser,
+} from '../../client';
+import { ChatMessage } from '../../message';
+import { ChatRoom } from '../../room';
+import AppChatUserPopover from '../../user-popover/user-popover.vue';
+import AppChatWindowOutputItemDetails from './AppChatWindowOutputItemDetails.vue';
+
+export interface ChatMessageEditEvent {
+	message: ChatMessage;
+}
+</script>
+
+<script lang="ts" setup>
+const props = defineProps({
+	message: {
+		type: Object as PropType<ChatMessage>,
+		required: true,
+	},
+	room: {
+		type: Object as PropType<ChatRoom>,
+		required: true,
+	},
+});
+
+const { message, room } = toRefs(props);
+
+const chatStore = inject<ChatStore>(ChatStoreKey)!;
+const { theme, isDark } = useThemeStore();
+
+const displayRules = new ContentRules({ maxMediaWidth: 400, maxMediaHeight: 300 });
+
+const root = ref<HTMLElement>();
+const itemWrapper = ref<HTMLElement>();
+
+const contentViewerBounds: ContentOwnerParentBounds = reactive({
+	width: ref(200),
+});
+
+const chat = computed(() => chatStore.chat!);
+
+// Use the form/page/user theme, or the default theme if none exist.
+const actualTheme = computed(() => theme.value ?? DefaultTheme);
+
+const isEditing = computed(() => chat.value.messageEditing === message.value);
+
+/** The background-color for chat items that are being edited. */
+const isEditingColor = computed(() => {
+	if (!isEditing.value) {
+		return undefined;
+	}
+
+	const highlight = '#' + actualTheme.value.highlight_;
+	const backlight = '#' + actualTheme.value.backlight_;
+
+	const tintColor = isDark.value ? highlight : backlight;
+	return transparentize(0.65, tintColor);
+});
+
+const editingState = computed(() => {
+	if (isEditing.value) {
+		return {
+			/** The text to display in the comment */
+			display: '(editing...)',
+			/** The tooltip to display on hover or tap */
+			tooltip: '',
+		};
+	}
+
+	if (message.value.edited_on) {
+		return {
+			display: '(edited)',
+			tooltip: formatDate(message.value.edited_on!, 'medium'),
+		};
+	}
+
+	return null;
+});
+
+const shouldShowMessageOptions = computed(() => canRemoveMessage.value || canEditMessage.value);
+
+const dataAnchorWidth = computed(() => {
+	const itemWidth = 40;
+	if (isEditing.value) {
+		return itemWidth;
+	}
+
+	let itemCount = 0;
+	if (canRemoveMessage.value) {
+		++itemCount;
+	}
+	if (canEditMessage.value) {
+		++itemCount;
+	}
+
+	return itemWidth * itemCount + (itemCount - 1);
```

Comment:
i don't understand this. Why is it adding itemCount - 1 at the end here? isn't that just gonna add like a pixel or two?

Reply:
the border/spacing between items is `1px`. rewrote this so it's a bit more clear without having to look at the stylus.